readthedocs/projects/views/public.py

```diff
@@ -122,59 +126,131 @@ def get_context_data(self, **kwargs):
         return context
 
 
-@never_cache
-def project_badge(request, project_slug):
-    """Return a sweet badge for the project."""
-    style = request.GET.get('style', 'flat')
-    if style not in (
+class ProjectBadgeView(View):
+
+    """
+    Return a sweet badge for the project.
+
+    Query parameters:
+
+    * ``version`` the version for the project (latest [default], stable, etc.)
+    * ``style`` the style of the badge (flat [default], plastic, etc.)
+    * ``token`` a project-specific token needed to access private versions
+    """
+
+    http_method_names = ['get', 'head', 'options']
+    STATUS_UNKNOWN = 'unknown'
+    STATUS_PASSING = 'passing'
+    STATUS_FAILING = 'failing'
+    STATUSES = (STATUS_FAILING, STATUS_PASSING, STATUS_UNKNOWN)
+
+    def get(self, request, project_slug, *args, **kwargs):
+        status = self.STATUS_UNKNOWN
+        token = request.GET.get('token')
+        version_slug = request.GET.get('version', LATEST)
+        version = None
+
+        if token:
+            version_to_check = Version.objects.filter(
+                project__slug=project_slug,
+                slug=version_slug,
+            ).first()
+            if version_to_check and self.verify_project_token(token, project_slug):
+                version = version_to_check
+        else:
+            version = Version.objects.public(request.user).filter(
+                project__slug=project_slug,
+                slug=version_slug,
+            ).first()
+
+        if version:
+            last_build = version.builds.filter(
+                type='html',
+                state='finished',
+            ).order_by('-date').first()
+            if last_build:
+                if last_build.success:
+                    status = self.STATUS_PASSING
+                else:
+                    status = self.STATUS_FAILING
+
+        return self.serve_badge(request, status)
+
+    def get_style(self, request):
+        style = request.GET.get('style', 'flat')
+        if style not in (
             'flat',
             'plastic',
             'flat-square',
             'for-the-badge',
             'social',
-    ):
-        style = 'flat'
-
-    # Get the local path to the badge files
-    badge_path = os.path.join(
-        os.path.dirname(__file__),
-        '..',
-        'static',
-        'projects',
-        'badges',
-        '%s-' + style + '.svg',
-    )
-
-    version_slug = request.GET.get('version', LATEST)
-    file_path = badge_path % 'unknown'
+        ):
+            style = 'flat'
+
+        return style
+
+    def serve_badge(self, request, status):
+        style = self.get_style(request)
+        if status not in self.STATUSES:
```

Comment:
We always pass a valid status here and looks like just a method used in this class.

Reply:
It's possible this is overkill as it isn't used anywhere else. However, it is nice to double check.